src/file/file.rpc.ts

```diff
@@ -19,52 +27,77 @@ const packageDefinition = protoLoader.loadSync(
+const protoDescriptor : grpc.GrpcObject = grpc.loadPackageDefinition(packageDefinition);
+const file_proto : any = protoDescriptor.file;
+
 export const serviceNames: string[] = ['', 'file.fileService'];
 export const healthCheckStatusMap = {
   '': HealthCheckResponse.ServingStatus.UNKNOWN,
   serviceName: HealthCheckResponse.ServingStatus.UNKNOWN
 };
 
-// Has the full package hierarchy
-const protoDescriptor = grpc.loadPackageDefinition(packageDefinition);
-const file_proto = protoDescriptor.file;
+
 
 /**
  * The FileServer class, containing all of the FileServer methods.
  */
 export class FileServer {
-  public server: any;
+
+  public server: grpc.Server;
   public grpcHealthCheck: GrpcHealthCheck;
+
   public constructor(port: string) {
     this.server = new grpc.Server();
+    this.addServices();
+    this.server.bind(`0.0.0.0:${port}`, grpc.ServerCredentials.createInsecure());
+  }
 
+  private addServices () {
     // Register the health service
     this.grpcHealthCheck = new GrpcHealthCheck(healthCheckStatusMap);
     this.server.addService(HealthService, this.grpcHealthCheck);
 
-    this.server.addService(file_proto.FileService.service, {
-      GenerateKey: this.generateKey,
-      CreateUpload: this.createUpload,
-      UpdateUploadID: this.updateUpload,
-      GetUploadByID: this.getUploadByID,
-      DeleteUploadByID: this.deleteUploadByID,
-      GetFileByID: this.getFileByID,
-      GetFileByKey: this.getFileByKey,
-      GetFilesByFolder: this.getFilesByFolder,
-      CreateFile: this.createFile,
-      DeleteFile: this.deleteFile,
-      IsAllowed: this.isAllowed,
-    });
+    const fileService = {
+      GenerateKey: this.wrapper(this.generateKey),
+      CreateUpload: this.wrapper(this.createUpload),
+      UpdateUploadID: this.wrapper(this.updateUpload),
+      GetUploadByID: this.wrapper(this.getUploadByID),
+      DeleteUploadByID: this.wrapper(this.deleteUploadByID),
+      GetFileByID: this.wrapper(this.getFileByID),
+      GetFileByKey: this.wrapper(this.getFileByKey),
+      GetFilesByFolder: this.wrapper(this.getFilesByFolder),
+      CreateFile: this.wrapper(this.createFile),
+      DeleteFile: this.wrapper(this.deleteFile),
+      IsAllowed: this.wrapper(this.isAllowed),
+    };
+
+    this.server.addService(file_proto.FileService.service, fileService);
+  }
+
+  private wrapper (func: Function) : any {
+    return async (call: grpc.ServerUnaryCall<Object>, callback: grpc.requestCallback<Object>) => {
+      try {
+        const traceparent = call.metadata.get('elastic-apm-traceparent');
+        const transOptions = (traceparent.length > 0) ? { childOf: traceparent[0].toString() } : {};
+        apm.startTransaction(`/file.FileService/${func.name}`, 'request', transOptions);
+        const res = await func(call, callback);
+        apm.endTransaction('OK');
+        callback(null, res);
+      } catch (err) {
+        const errName = err.code ? statusToString(err.code) : 'UNKNOWN';
```

Comment:
instead of handle error like this, write a function that validated and cast the error to grpc error then pass the name and err to apm.endTransaction and callback